modules/javafx.graphics/src/main/resources/com/sun/prism/es2/glsl/main2Lights.frag

```diff
@@ -60,13 +60,74 @@ struct Light {
     vec3 color;
     vec3 attn;
     float range;
+    vec3 dir;
+    float cosOuter;
+    float denom; // cosInner - cosOuter
+    float falloff;
 };
 
 uniform vec3 ambientColor;
 uniform Light lights[3];
 
 varying vec3 eyePos;
 varying vec4 lightTangentSpacePositions[3];
+varying vec4 lightTangentSpaceDirections[3];
+
+// Because pow(0, 0) is undefined (https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/pow.xhtml),
+// we need special treatment for falloff == 0 cases
+float computeSpotlightFactor(vec3 l, vec3 lightDir, float cosOuter, float denom, float falloff) {
+    if (falloff == 0.0 && cosOuter == -1.0) { // point light optimization (cosOuter == -1 is outerAngle == 180)
+        return 1.0;
+    }
+    float cosAngle = dot(normalize(-lightDir), l);
+    float cutoff = cosAngle - cosOuter;
+    if (falloff != 0.0) {
+        return pow(clamp(cutoff / denom, 0.0, 1.0), falloff);
+    }
+    return cutoff >= 0.0 ? 1.0 : 0.0;
+}
+
+float computeSpotlightFactor2(vec3 l, vec3 lightDir, float cosOuter, float denom, float falloff) {
+    if (falloff != 0.0) {
+        float cosAngle = dot(normalize(-lightDir), l);
+        float cutoff = cosAngle - cosOuter;
+        return pow(clamp(cutoff / denom, 0.0, 1.0), falloff);
+    }
+    if (cosOuter == -1.0) {  // point light optimization (cosOuter == -1 is outerAngle == 180)
+        return 1.0;
+    }
+    float cosAngle = dot(normalize(-lightDir), l);
+    float cutoff = cosAngle - cosOuter;
+    return cutoff >= 0.0 ? 1.0 : 0.0;
+}
+
+float computeSpotlightFactor3(vec3 l, vec3 lightDir, float cosOuter, float denom, float falloff) {
+    float cosAngle = dot(normalize(-lightDir), l);
+    float cutoff = cosAngle - cosOuter;
+    if (falloff != 0.0) {
+        return pow(clamp(cutoff / denom, 0.0, 1.0), falloff);
+    }
+    return cutoff >= 0.0 ? 1.0 : 0.0;
+}
+
+void computeLight(int i, vec3 n, vec3 refl, float specPower, out vec3 d, out vec3 s) {
```

Comment:
The storage modifier for `d` and `s` must be changed to `inout` instead of `out`.